### PR TITLE
[5.2] atum light mode - Fix modal window link color

### PR DESCRIPTION
**Pull request by RickR2H**
Pull Request for Issue # .

### Summary of Changes
With the browser in dark modus, In the backend (Atum template) the links have the wrong color when the template is toggled to the light modus. The reason is that in the component.php only the dark color scheme colors are loaded if the browser is is dark mode. My fix will make sure that the correct colors are used, depending or the color scheme selected.

In dark modus the default hover color in the modal is defaulted to white, which is not correct. This PR fixes also the hover color which is now set to a lighter shade of the default link color.

### Testing Instructions
1. Set the browser you are using in dark mode.
2. In the backend from Joomla select from the User Menu in the top right the link: Light/Dark Mode, and switch the template to light modus.
3. Open a modal. This can be done by opening an article and selecting the Created By button in the Publishing tab.
image
4. The link color has a light blue color which is not accessible.
5. Switch Joomla to Dark mode.
6. The link hover color is white. 

#### Apply the PR
1. Switch to light modus.
2. The link color is now the correct color and readable.
3. Switch to dark mode.
7. The link hover color is now light blue.

#### The link is too light which is a accessibility issue.
image

#### In dark mode the hover color white
image

#### The link has the correct color.
image

#### In dark mode the hover color light blue
image

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on 61500560b6faf9d79d8c2e400cdaabc0363f118f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44176">issues.joomla.org/tracker/joomla-cms/44176</a>.</sub>

**Comment by brianteeman** [comment]
Please also check #43708

**Comment by richard67** [comment]
@RickR2H Adding the `// phpcs:disable PSR1.Files.SideEffects` hasn't helped. I think the issue is not the `defined('_JEXEC') or die;` but the function definition here: https://github.com/joomla/joomla-cms/pull/44176/files#diff-3538b524e6cf4f442f5ca0631475c0ef2b867edbcb23a24b5deb2248c428e252R35-R42 . Maybe it needs to wrap that with `// phpcs:disable PSR1.Files.SideEffects` and `// phpcs:enable PSR1.Files.SideEffects`?

**Comment by fgsw** [comment]
@RickR2H Can you add ```[5.2]``` to the title so user don't have to open the Pull Request to know which branche it is for?

**Comment by richard67** [comment]
I‘ve restored the previous human test result as the changes after that were only hints for PHPCS.

**Comment by richard67** [comment]
@RickR2H Please don’t do unnecessary branch Updates if there are no conflicts as the branch update invalidates the human test counter and I cannot watch your PR all day long to restore the tests.

File: administrator/templates/atum/component.php [hunk 1]
```diff
@@ -26,6 +26,20 @@
 
 $linkColorDark = $this->params->get('link-color-dark', '#6fbfdb');
 list($rd, $gd, $bd) = sscanf($linkColorDark, "#%02x%02x%02x");
+list($lighterRd, $lighterGd, $lighterBd) = adjustColorLightness($rd, $gd, $bd, 10);
+
+$linkColorDarkHvr = sprintf("%d, %d, %d", $lighterRd, $lighterGd, $lighterBd);
+
+// phpcs:disable PSR1.Files.SideEffects
```
Comment (dgrammatiko) [inline administrator/templates/atum/component.php]: If you move the function before line 29 you probably don't need to disable the checker. You're using a function before declaring it (JS has hoisting but I don't think PHP does thus my comment)

FWIW that was also the warning message: 

```
1 | WARNING | A file should declare new symbols (classes, functions,
   |         | constants, etc.) and cause no other side effects, or it should
   |         | execute logic with side effects, but should not do both. The
   |         | first symbol is defined on line 33 and the first side effect is
   |         | on line 11.
```

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

File: administrator/templates/atum/component.php (hunk 1 above)
Comment (RickR2H) [reply administrator/templates/atum/component.php]: Will try that. Thanks!

**Comment by RickR2H** [comment]
@dgrammatiko @richard67 I can't get drone to behave... Any tips?

**Comment by richard67** [comment]
> @dgrammatiko @richard67 I can't get drone to behave... Any tips?

With the previous version with my proposal it was ok. So I suggest to wrap the method again into the `// phpcs:disable PSR1.Files.SideEffects` and `// phpcs:enable PSR1.Files.SideEffects`.

File: administrator/templates/atum/component.php [hunk 2]
```diff
@@ -37,6 +30,17 @@ function adjustColorLightness($r, $g, $b, $percent)
...
+function adjustColorLightness($r, $g, $b, $percent)
+{
+    $adjust = function ($color) use ($percent) {
+        $newColor = $color + ($color * $percent / 100);
+        return min(max(0, $newColor), 255);
+    };
+    return [$adjust($r), $adjust($g), $adjust($b)];
+}
+// phpcs:enable PSR1.Files.SideEffects
+
```
Comment (richard67) [inline administrator/templates/atum/component.php:37]: I wouldn't move that function back down to after the place where it's called first time like you did with your latest commit.

Comment (dgrammatiko) [reply administrator/templates/atum/component.php:37]: PHP is fine assigning a function to a variable? if so try that

Comment (RickR2H) [reply administrator/templates/atum/component.php:37]: Looks like drone is okay now. If you all don't mind, I'll like to keep it as is for now.

**Comment by fgsw** [comment]
I have tested this item :white_check_mark: successfully on c21bf325d7b6ce7e8d0f8dcaa573402b61bbebea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44176">issues.joomla.org/tracker/joomla-cms/44176</a>.</sub>

File: administrator/templates/atum/component.php [hunk 3]
```diff
@@ -8,7 +8,9 @@
  * @license     GNU General Public License version 2 or later; see LICENSE.txt
  */
 
+// phpcs:disable PSR1.Files.SideEffects
 defined('_JEXEC') or die;
+// phpcs:enable PSR1.Files.SideEffects
```
Comment (richard67) [inline administrator/templates/atum/component.php]:
```suggestion
defined('_JEXEC') or die;
```
PHPCS would also be happy without that.

File: administrator/templates/atum/component.php [hunk 4]
```diff
@@ -26,6 +28,20 @@
...
+{
+    $adjust = function ($color) use ($percent) {
+        $newColor = $color + ($color * $percent / 100);
+        return min(max(0, $newColor), 255);
+    };
+    return [$adjust($r), $adjust($g), $adjust($b)];
+}
+// phpcs:enable PSR1.Files.SideEffects
+
+$linkColorDarkHvr = sprintf("%d, %d, %d", $lighterRd, $lighterGd, $lighterBd);
```
Comment (richard67) [inline administrator/templates/atum/component.php]:
```suggestion

$adjustColorLightness = function ($r, $g, $b, $percent) {
    $adjust = function ($color) use ($percent) {
        $newColor = $color + ($color * $percent / 100);
        return min(max(0, $newColor), 255);
    };
    return [$adjust($r), $adjust($g), $adjust($b)];
};

list($lighterRd, $lighterGd, $lighterBd) = $adjustColorLightness($rd, $gd, $bd, 10);
$linkColorDarkHvr = sprintf("%d, %d, %d", $lighterRd, $lighterGd, $lighterBd);
```
In this way PHPCS would accept it without the `// phpcs:disable PSR1.Files.SideEffects`.

Comment (richard67) [reply administrator/templates/atum/component.php]: P.S.: I've tested my 2 suggestions here and they work, and both our PHPCS checkers are happy with it, too.

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on c21bf325d7b6ce7e8d0f8dcaa573402b61bbebea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44176">issues.joomla.org/tracker/joomla-cms/44176</a>.</sub>

**Comment by richard67** [comment]
@RickR2H Could you check my 2 suggestions for how to get rid of these PHPCS hints? It would not need new human tests if you apply them.

File: administrator/templates/atum/component.php (hunk 3 above)
Comment (RickR2H) [reply administrator/templates/atum/component.php]: Removed the tags

**Comment by C-Lodder** [comment]
Why use PHP to manipulate CSS colours, when this can be acheived in native CSS with 1 line of code?

**Comment by dgrammatiko** [comment]
> Why use PHP to manipulate CSS colours, when this can be acheived in native CSS with 1 line of code?

Because BS is using a css var of type `number, number, number` : 
https://github.com/twbs/bootstrap/blob/fee9dc2438736a02405412e10ea8445dd9aeac1e/dist/css/bootstrap.css#L40-L47

**Comment by richard67** [comment]
I've restored the previous human test results in the issue tracker as the commits which have invalidated the test count were only related to making PHPCS happy. In addition I have tested that the PR still works with the last change to an anonymous function.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44176">issues.joomla.org/tracker/joomla-cms/44176</a>.</sub>

**Comment by C-Lodder** [comment]
@dgrammatiko https://jsfiddle.net/e5arm7db ?

**Comment by dgrammatiko** [comment]
July 2024 might be too soon:
https://caniuse.com/css-relative-colors
Screenshot 2024-10-08 at 5 13 54 PM

Anyways if the maintainers are ok with it, go for it, it's not my call. I just pointed that the BS requires a weird value and that the relative color from is a new addition 🤷‍♂️

**Comment by richard67** [comment]
@RickR2H To be honest, to me the hover colour seems to be inconsistent now when comparing dark mode and light mode.

In light mode it is still black with your PR, so when changing to dark mode I would expect it to be white.

Or vice versa, when coming from dark mode where it is light blue with your PR, and changing to light mode, I would expect it to be some other kind of blue but not black.

But maybe that's just me.

**Comment by RickR2H** [comment]
I guess there will be a better solution. I have no clue why the values have to calculated. We can make then fixed variables which is a better option. Probably is has to do with the backend colors...

**Comment by dgrammatiko** [comment]
> I have no clue why the values have to calculated.

The blame here goes to Bootstrap, they store the rgb values as a coma separated val: `--bs-dark-rgb: 33, 37, 41;` https://github.com/twbs/bootstrap/blob/fee9dc2438736a02405412e10ea8445dd9aeac1e/dist/css/bootstrap.css#L40-L47

> Probably is has to do with the backend colors...

Nope, search the bs css and you'll see how these values are used

> I guess there will be a better solution.

What Charlie suggested but the browser support is only few months...

**Comment by wilsonge** [comment]
This is consistent with index.php so I'd merge this if it fixes the bug - then at least if we ever change to @C-Lodder 's way in the future (j6 or whatever) then we can do it consistently. Only thing I'd suggest is that we have this the error_login.php and error_full.php too so we have the same set of css variables in all 3 files.

**Comment by pe7er** [comment]
Thanks @RickR2H !

**Comment by dgrammatiko** [comment]
> Only thing I'd suggest is that we have this the error_login.php and error_full.php too so we have the same set of css variables in all 3 files.

@RickR2H please do another PR with the same code covering the other 2 files @wilsonge mentioned. Thanks

**Pull request merged**